### PR TITLE
fix: centralize download path sanitization and persist target paths

### LINKED ISSUE #175: Centralize download path sanitization and persist exact target paths for resume/cleanup
## Investigative sweep finding

Download path handling is inconsistent across initial start, restart/resume, torrent, and cleanup paths. Some flows sanitize names, others do not; paused-download cleanup also calculates the wrong folder.

### Evidence

- Direct initial download sanitizes `result.name` and `file.name`: `application/src/frontend/lib/downloads/services/DirectService.ts:11-17` and `:40-45`.
- Torrent initial download uses unsanitized `result.name` to build `downloadPath`: `application/src/frontend/lib/downloads/services/TorrentService.ts:31` and stores the unsanitized path at `:60` / `:88`.
- Direct restart/resume reconstructs paths with unsanitized `download.name` and `file.name`: `application/src/frontend/lib/downloads/restart.ts:37-50`.
- Torrent restart sanitizes only fallback filename, then uses unsanitized `download.name` in the containing directory: `application/src/frontend/lib/downloads/restart.ts:128-156`.
- `deleteDownloadedItems()` trims `downloadPath` and then takes its parent directory before appending file names: `application/src/frontend/lib/downloads/persistence.ts:291-318`. For a normal `downloads/Game/` path, this becomes `downloads/<file>` instead of `downloads/Game/<file>`.

### User-visible failure modes / edge cases

- Add-ons returning names with `/`, `\`, `..`, `:`, or Windows-reserved characters can write into unexpected folders or fail only for torrent/resume flows.
- A direct download started with sanitized names can resume into a different unsanitized path, duplicating/restarting instead of continuing the existing partial file.
- Paused direct-download cleanup may fail to delete partial files because it points one directory too high.
- Multi-part files with nested paths or duplicate sanitized basenames can collide.
- Different platforms will behave differently because Windows path-invalid characters are only handled in some code paths.

### Suggested fix

- Centralize a single `sanitizePathSegment()` / `safeDownloadPath()` helper used by direct, torrent, Debrid, restart/resume, and cleanup.
- Persist the exact resolved target paths used by the backend instead of reconstructing from display names.
- Make cleanup delete exactly the persisted target paths, not paths recomputed from `downloadPath` + file name.
- Add tests/cases for Windows-reserved characters, path traversal strings, duplicate basenames, and pause/resume cleanup.

Related existing issue: #162 (direct download refactor).

### PR DESCRIPTION
**Pull request by nat-openclaw**
## Summary
- Add shared `paths.ts` helpers (`sanitizePathSegment`, `safeDownloadPath`, `getPersistedFilePaths`) used by all download start flows
- Persist exact per-file `path` values in download state so resume and cleanup use the same targets the backend wrote
- Fix `deleteDownloadedItems()` deleting from the wrong parent directory on paused direct downloads
- Add unit tests for sanitization, traversal strings, duplicate basenames, and cleanup path resolution

Closes #175

## Test plan
- [x] `bun test src/frontend/lib/downloads/paths.test.ts` (12 tests)
- [x] `bun run typecheck` in `application/` (0 errors)
- [ ] Manual: start a direct download with special chars in name, pause, delete — file removed from correct folder
- [ ] Manual: pause/resume direct download continues same partial file (no duplicate path)

Made with [Cursor](https://cursor.com)

### CODE REVIEW

**Comment by Nat3z** [comment]
@nat-openclaw remove the test file and get up to date with main

**Comment by nat-openclaw** [comment]
<details>
<summary>▶️ Actions Performed</summary>

- [x] Delete paths.test.ts and clean up package.json/tsconfig — Removed paths.test.ts, test script, and tsconfig exclude
- [x] Merge origin/main into PR branch — Clean merge of 3 main commits (updater fixes)
- [x] Run bun run typecheck in application/ — typecheck passed after build:all
- [x] Commit and push to PR branch — Pushed c6a3749 to PR branch

</details>

**Comment by nat-openclaw** [comment]
Removed `application/src/frontend/lib/downloads/paths.test.ts`, cleaned up the related test script/exclude config, and merged `main` into the branch so it’s up to date.

Also picked up the latest updater/startup fixes from `main`. I ran `bun run build:all && bun run typecheck` in `application/`, and both passed.

Pushed as `c6a3749`.

File: application/src/frontend/lib/downloads/paths.ts [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+  const base = baseDir.replace(/[/\\]+$/, '');
+  const folder = sanitizePathSegment(folderName);
+  if (fileName === undefined) {
+    return `${base}/${folder}/`;
+  }
+  const file = sanitizePathSegment(fileName);
+  return `${base}/${folder}/${file}`;
+}
+
+/** Ensures unique basenames so multi-file torrents do not overwrite each other. */
```
Comment (Nat3z) [inline application/src/frontend/lib/downloads/paths.ts:31]:
```suggestion

File: application/src/frontend/store.ts [hunk 2]
```diff
@@ -31,6 +31,8 @@ export type DownloadStatusAndInfo = SearchResult & {
   downloadPath: string;
   files: {
     name: string;
+    /** Exact resolved target path used by the download backend. */
+    path?: string;
```
Comment (Nat3z) [inline application/src/frontend/store.ts:35]: and so what exactly is the purpose of including the path? why cant we just like infer it from the system?

**Review by Nat3z — COMMENTED** [review]
@nat-openclaw please answer the questions before i merge. /ask

**Comment by nat-openclaw** [comment]
Asked and answered — no code changes were made for `/ask`.

**Comment by nat-openclaw** [comment]
Yep — the `path` is there because `downloadPath` is only the destination folder, not the actual file path.

We can’t reliably infer the real file location from just the system/name because:
- start flows sanitize names before writing
- duplicate names can get deduped (`foo.iso` → `foo_2.iso`)
- different download services have historically built paths a bit differently
- cleanup/resume needs to hit the exact same on-disk target the backend originally wrote

So persisting `files[].path` makes pause/delete/resume operate on the same file the downloader created, instead of reconstructing a “best guess” and occasionally yeeting the wrong folder. Very classic filesystem goblin behavior.

On the ` ```suggestion ` for `dedupeFileNames`: that looks empty/incomplete, so I’m not sure what change you wanted there. If you meant a specific edit, paste the actual suggestion and I’ll answer it directly.

**Review by Nat3z — APPROVED** [review]
lgtm
